include/revng/ADT/GenericGraph.h

```diff
@@ -353,11 +370,557 @@ class BidirectionalNode
+      [](NonOwningEdge const &E) -> DerivedType const *const & {
+      return E.Neighbor;
+    };
+  };
+
+  using SuccessorEdgeIterator = revng::
+    mapped_iterator<OwningEdge *, typename SuccessorFilters::Types::ToView>;
+  using ConstSuccessorEdgeIterator = revng::mapped_iterator<
+    OwningEdge *,
+    typename SuccessorFilters::Types::ToConstView>;
+  using SuccessorIterator = revng::
+    mapped_iterator<OwningEdge *, typename SuccessorFilters::Types::ToNeighbor>;
+  using ConstSuccessorIterator = revng::mapped_iterator<
+    OwningEdge *,
+    typename SuccessorFilters::Types::ToConstNeighbor>;
+
+  using PredecessorEdgeIterator = revng::mapped_iterator<
+    NonOwningEdge *,
+    typename PredecessorFilters::Types::ToView>;
+  using ConstPredecessorEdgeIterator = revng::mapped_iterator<
+    NonOwningEdge *,
+    typename PredecessorFilters::Types::ToConstView>;
+  using PredecessorIterator = revng::mapped_iterator<
+    NonOwningEdge *,
+    typename PredecessorFilters::Types::ToNeighbor>;
+  using ConstPredecessorIterator = revng::mapped_iterator<
+    NonOwningEdge *,
+    typename PredecessorFilters::Types::ToConstNeighbor>;
+
+public:
+  llvm::iterator_range<SuccessorEdgeIterator> successor_edges() {
+    auto Range = llvm::make_range(Successors.begin(), Successors.end());
+    return revng::map_range(Range, SuccessorFilters::ToView);
+  }
+  llvm::iterator_range<ConstSuccessorEdgeIterator> successor_edges() const {
+    auto Range = llvm::make_range(Successors.begin(), Successors.end());
+    return revng::map_range(Range, SuccessorFilters::ToConstView);
+  }
+  llvm::iterator_range<SuccessorIterator> successors() {
+    auto Range = llvm::make_range(Successors.begin(), Successors.end());
+    return revng::map_range(Range, SuccessorFilters::ToNeighbor);
+  }
+  llvm::iterator_range<ConstSuccessorIterator> successors() const {
+    auto Range = llvm::make_range(Successors.begin(), Successors.end());
+    return revng::map_range(Range, SuccessorFilters::ToConstNeighbor);
+  }
+
+  llvm::iterator_range<PredecessorEdgeIterator> predecessor_edges() {
+    auto Range = llvm::make_range(Predecessors.begin(), Predecessors.end());
+    return revng::map_range(Range, PredecessorFilters::ToView);
+  }
+  llvm::iterator_range<ConstPredecessorEdgeIterator> predecessor_edges() const {
+    auto Range = llvm::make_range(Predecessors.begin(), Predecessors.end());
+    return revng::map_range(Range, PredecessorFilters::ToConstView);
+  }
+  llvm::iterator_range<PredecessorIterator> predecessors() {
+    auto Range = llvm::make_range(Predecessors.begin(), Predecessors.end());
+    return revng::map_range(Range, PredecessorFilters::ToNeighbor);
+  }
+  llvm::iterator_range<ConstPredecessorIterator> predecessors() const {
+    auto Range = llvm::make_range(Predecessors.begin(), Predecessors.end());
+    return revng::map_range(Range, PredecessorFilters::ToConstNeighbor);
+  }
+
+protected:
```

Comment:
It's just that if I ever want to extend this node by inheriting, I'd want to use these `find` function instead of the public ones, mostly to avoid extra iteration conversions.
These use `real` iterators as opposed to `mapped` public ones.

Reply:
I think that if we ever need to extend this class we can relax the `private`.
In general, let's avoid future hypothetical use cases. This often lead to code harder to maintain with only hypothetical future benefits that oftentimes never take place.

Not a big deal here, but it's a general rule.

Reply:
Ok, I got you. `private` also works.

Reply:
Done.